mgplvm/rdist/GPbase.py

```diff
@@ -0,0 +1,177 @@
+import torch
+import numpy as np
+from torch import nn, Tensor
+from torch.distributions.multivariate_normal import MultivariateNormal
+from ..utils import softplus, inv_softplus
+from ..manifolds.base import Manifold
+from .common import Rdist
+from typing import Optional
+from ..fast_utils.toeplitz import sym_toeplitz_matmul
+
+
+class GPbase(Rdist):
+    name = "GPbase"
+
+    def __init__(self,
+                 manif: Manifold,
+                 m: int,
+                 n_samples: int,
+                 ts: torch.Tensor,
+                 _scale=0.9,
+                 ell=None):
+        """
+        Parameters
+        ----------
+        manif: Manifold
+            manifold of ReLie
+        m : int
+            number of conditions/timepoints
+        n_samples: int
+            number of samples
+        ts: Tensor
+            input timepoints for each sample (n_samples x 1 x m)
+        mu : Optional[np.ndarray]
+            initialization of the vartiational means (m x d2)
+            
+        Notes
+        -----
+        Our GP has prior N(0, K)
+        We parameterize our posterior as N(K2 v, K2 I^2 K2)
+        where K2 K2 = K and I(s) is some inner matrix which can take different forms.
+        s is a vector of scale parameters for each time point.
+        
+        """
+
+        super(GPbase, self).__init__(manif, 1)  #kmax = 1
+
+        self.manif = manif
+        self.d = manif.d
+        self.m = m
+
+        #initialize GP mean parameters
+        nu = torch.randn((n_samples, self.d, m)) * 0.01
+        self._nu = nn.Parameter(data=nu, requires_grad=True)  #m in the notes
+
+        #initialize covariance parameters
+        _scale = torch.ones(n_samples, self.d, m) * _scale  #n_diag x T
+        self._scale = nn.Parameter(data=inv_softplus(_scale),
+                                   requires_grad=True)
+
+        #initialize length scale
+        ell = (torch.max(ts) - torch.min(ts)) / 20 if ell is None else ell
+        _ell = torch.ones(1, self.d, 1) * ell
+        self._ell = nn.Parameter(data=inv_softplus(_ell), requires_grad=True)
+
+        #pre-compute time differences (only need one row for the toeplitz stuff)
+        self.ts = ts
+        dts_sq = torch.square(ts - ts[..., :1])  #(n_samples x 1 x m)
+        #sum over _input_ dimension, add an axis for _output_ dimension
+        dts_sq = dts_sq.sum(-2)[:, None, ...]  #(n_samples x 1 x m)
+        self.dts_sq = nn.Parameter(data=dts_sq, requires_grad=False)
+
+    @property
+    def scale(self) -> torch.Tensor:
+        return softplus(self._scale)
+
+    @property
+    def nu(self) -> torch.Tensor:
+        return self._nu
+
+    @property
+    def ell(self) -> torch.Tensor:
+        return softplus(self._ell)
+
+    @property
+    def prms(self):
+        return self.nu, self.scale, self.ell
+
+    def K_half(self, sample_idxs):
```

Comment:
Is there a reference for how K_half is calculated based on the prior kernel? If not, is it worth me going through the derivation and typing it up to add to the paper?

Reply:
(to verify it and make sure I agree with what you've done, though it seems to work empirically)

Reply:
I have some notes lying around somewhere that we can start from but they're not very clean. There's actually also a dt in there which I should add.

Reply:
should the dt be in this PR or we can add it later? isn't everything parameterized in terms of dt=1 anyhow in the package right now?

Reply:
I'll add it in the initialization here